### ISSUE TITLE
Photos fail to index after import, empty file error

### ISSUE THREAD

**Issue by aaliddell** [photoprism/photoprism#2516]
#### 1. What is not working as documented?

Photos are being added to PhotoPrism via iOS PhotoSync app to WebDAV. When import runs, it first moves the .heic files and then fails with error saying the file it just moved is empty. Looking at the files on disk, they are not empty. If I then run a manual reindex, the files are correctly picked up without issue. The library is stored on a network mounted drive, which may be important here if there is a race between move and read.

#### 2. How can we reproduce it?

1. Sync .heic file from iOS to WebDAV
2. Wait for auto import or run manual import from UI
3. Observe that files are moved but fail to index with empty file error
4. Run manual reindex and observe files are now indexed

- Imports are usually batches of about 10 photos
- Library is about 10000 items

#### 3. What behavior do you expect?

Files should import correctly on first pass without manual reindex

#### 4. What could be the cause of your problem?

Race condition between the file move and the subsequent indexing? The file may not have finished moving before the index looks and finds an empty file. Being on a network drive may be exposing a flaw here due to RTT for disk access. This bug may still be present on local drives, but the window for it to occur may just be much smaller.

Notably, this worked fine in previous versions, but I cannot give you an exact point this started failing. Within last few months as approximate guess.

Error is coming here:
https://github.com/photoprism/photoprism/blob/99a99a0868e25282db62ac3d62fe8237d9f3e71d/internal/photoprism/import_worker.go#L121-L126

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

Snapshot of logs coming shortly

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: AMD64 220629-5d7448d2

(b) Database Type & Version: MariaDB

(c) Operating System Types & Versions: Linux

(d) Browser Types & Versions: N/A

(e) Ad Blockers, Browser Plugins, and/or Firewall Software? N/A

#### 7. On what kind of device is PhotoPrism installed?

(a) Device / Processor Type: Intel Core i5
(b) Physical Memory & Swap Space in GB: 12 GB
(c) Storage Type: Database on local SSD, Library on remote SMB (import dir and library both on same mountpoint)
(d) Anything else that might be helpful to know? N/A

#### 8. Do you use a Reverse Proxy, Firewall, VPN, or CDN?

N/A

**Comment by lastzero** [comment]
Seems to be a permissions issue. If a file can't be read, it appears empty although it is not.

**Comment by aaliddell** [comment]
Nothing is changing between the 'empty file' error and the subsequent successful index though. So if there was a permissions error, I would expect either the move to fail and the subsequent index to fail.

Promised logs below tracing a single photo:

```
time="2022-07-14T07:00:11Z" level=info msg="http: starting web server at 0.0.0.0:2342"
...
time="2022-07-14T13:37:43Z" level=info msg="webdav: PUT /import/Phone/Recents/IMG_0285.HEIC"
...
time="2022-07-14T13:38:11Z" level=info msg="copying files from import"
time="2022-07-14T13:38:11Z" level=info msg="classify: loading nasnet"
time="2022-07-14T13:38:15Z" level=info msg="classify: loading labels from labels.txt"
time="2022-07-14T13:38:15Z" level=info msg="import: found no .ppignore file"
time="2022-07-14T13:38:15Z" level=info msg="media: IMG_0285.HEIC was taken at 2022-06-30 17:35:39 +0000 UTC (meta)"
time="2022-07-14T13:38:16Z" level=info msg="import: moving main heif file Phone/Recents/IMG_0285.HEIC to 2022/06/20220630_173539_6008438B.heic"
time="2022-07-14T13:38:16Z" level=error msg="import: 2022/06/20220630_173539_6008438B.heic is empty in 2022/06/20220630_173539_6008438B.heic"
...
time="2022-07-14T13:38:38Z" level=info msg="import completed in 26 s"
```

Start of manual index:
```
time="2022-07-14T13:59:56Z" level=info msg="indexing originals..."
time="2022-07-14T13:59:56Z" level=info msg="index: found no .ppignore file"
...
time="2022-07-14T14:05:56Z" level=info msg="convert: converting 20220630_173539_6008438B.heic to 20220630_173539_6008438B.heic.jpg (heif-convert)"
time="2022-07-14T14:05:58Z" level=info msg="convert: 20220630_173539_6008438B.heic.jpg created in 1.162203993s (heif-convert)"
time="2022-07-14T14:05:59Z" level=info msg="media: created 11 thumbnails for 20220630_173539_6008438B [1.676261021s]"
time="2022-07-14T14:05:59Z" level=info msg="index: added main heif file 2022/06/20220630_173539_6008438B.heic"
time="2022-07-14T14:06:00Z" level=info msg="index: matched 1 label with 20220630_173539_6008438B.heic.jpg [343.911628ms]"
time="2022-07-14T14:06:00Z" level=info msg="index: stacked related jpg file 2022/06/20220630_173539_6008438B.heic.jpg"
...
time="2022-07-14T14:08:04Z" level=info msg="indexing completed in 488 s"
time="2022-07-14T14:15:29Z" level=info msg="metadata: updated 12 photos"
time="2022-07-14T16:30:30Z" level=info msg="metadata: updated 3 photos"
```

**Comment by lastzero** [comment]
Error reporting must be improved, didn't get to it yet due to the amount of other work we have. The most time consuming part is to reproduce it as everything works under normal conditions.

**Comment by lastzero** [comment]
It could also be a sync issue since you mentioned a network file system. I remember someone had that as well. Could you try to use NFS instead of SMB or change the mount options (sync, async,...) if there are any?

**Comment by aaliddell** [comment]
I’ll see if I can reduce the steps required to a minimal case. My bets are on some weird network mount behaviour, although I’ve dug through the cifs mount options and it doesn’t seem I’ve got any weird options set.

Looking at the code, it seems like the issue is that stat is indeed returning 0 bytes. However, if I manually make a file, move it and stat it on the mount in quick succession, I get the correct nonzero output.

I’ll see if I can strace PhotoPrism whilst it’s doing an import too; that’ll tell us if the fs is really saying the file is zero bytes or if the error is at a higher layer.

**Comment by aaliddell** [comment]
Well this is odd.

If I enable debug log level, I see that it is switching from move to copy/delete due to the source and destination of the move being on different mounts. This is an artifact of being under docker, where `.../import` and `.../originals` are two separate mounts on the container, despite being the same underlying network mount on the host. This is not unexpected:

```
time="2022-07-14T19:50:38Z" level=info msg="import: moving main heif file Phone/Recents/IMG_0337.HEIC to 2022/07/20220714_194955_4035A3B4.heic"
time="2022-07-14T19:50:38Z" level=debug msg="failed renaming file, fallback to copy and delete: rename /photoprism/import/Phone/Recents/IMG_0337.HEIC /photoprism/originals/2022/07/20220714_194955_4035A3B4.heic: invalid cross-device link"
time="2022-07-14T19:50:38Z" level=error msg="import: 2022/07/20220714_194955_4035A3B4.heic is empty in 2022/07/20220714_194955_4035A3B4.heic
```

Then if I watch strace output during this, I see it indeed tries a `renameat`, then does a copy and unlink when that fails. The subsequent `newfsstatat` gets `st_size=0`, so the filesystem really thinks the size is zero after the write.

```
[pid 153549] 19:50:38 openat(AT_FDCWD, "/photoprism/import/Phone/Recents/IMG_0337.HEIC", O_RDONLY|O_CLOEXEC) = 8 <0.000905>                                                                                                                                                   
[pid 153549] 19:50:38 close(8)          = 0 <0.000517>                                                                                                                                                                                                                        
[pid 153580] 19:50:38 newfstatat(AT_FDCWD, "/photoprism/originals/2022/07/20220714_194955_4035A3B4.heic", 0xc001fa3d78, 0) = -1 ENOENT (No such file or directory) <0.001732>                                                                                                 
[pid 153580] 19:50:38 newfstatat(AT_FDCWD, "/photoprism/originals/2022/07", {st_mode=S_IFDIR|0755, st_size=0, ...}, 0) = 0 <0.000065>                                                                                                                                         
[pid 153580] 19:50:38 newfstatat(AT_FDCWD, "/photoprism/originals/2022/07", {st_mode=S_IFDIR|0755, st_size=0, ...}, 0) = 0 <0.000093>                                                                                                                                         
[pid 153580] 19:50:38 newfstatat(AT_FDCWD, "/photoprism/originals/2022/07/20220714_194955_4035A3B4.heic", 0xc006b28038, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory) <0.000110>                                                                               
[pid 153580] 19:50:38 renameat(AT_FDCWD, "/photoprism/import/Phone/Recents/IMG_0337.HEIC", AT_FDCWD, "/photoprism/originals/2022/07/20220714_194955_4035A3B4.heic") = -1 EXDEV (Invalid cross-device link) <0.000091>                                                         
[pid 153580] 19:50:38 newfstatat(AT_FDCWD, "/photoprism/originals/2022/07", {st_mode=S_IFDIR|0755, st_size=0, ...}, 0) = 0 <0.000024>
[pid 153580] 19:50:38 openat(AT_FDCWD, "/photoprism/import/Phone/Recents/IMG_0337.HEIC", O_RDONLY|O_CLOEXEC) = 8 <0.000840>
[pid 153580] 19:50:38 openat(AT_FDCWD, "/photoprism/originals/2022/07/20220714_194955_4035A3B4.heic", O_RDWR|O_CREAT|O_CLOEXEC, 0777) = 9 <0.001460>
[pid 153580] 19:50:38 close(9)          = 0 <0.004047>
[pid 153580] 19:50:38 close(8)          = 0 <0.000716>
[pid 153580] 19:50:38 unlinkat(AT_FDCWD, "/photoprism/import/Phone/Recents/IMG_0337.HEIC", 0) = 0 <0.000866>
[pid 153580] 19:50:38 newfstatat(AT_FDCWD, "/photoprism/originals/2022/07/20220714_194955_4035A3B4.heic", {st_mode=S_IFREG|0755, st_size=0, ...}, 0) = 0 <0.000077>
```

So next I watch the SMB traffic with WireShark and whilst I can see the copy happen, it never reloads the file info afterwards, suggesting the local machine is keeping an old inode cache entry around momentarily that says the file size is still zero. Setting cifs mount option `cache=none` doesn't help. So either this is a filesystem consistency bug or this is just one of those things that a filesystem is permitted to do without an explicit `sync` call; probably the latter. Weirdly though I cannot reproduce this behaviour with any other program despite following the exact same logical flow of create -> copy -> stat.

Consider me confused... :confused: 

However, circling all the way back around to PhotoPrism: why bother `stat`ing the file and checking for zero length? The subsequent steps in ` NewMediaFile` immediately open and read the file, so would fail anyway on a legitimately empty file without the zero check. However, in this case the invalid inode cache won't apply during a read, so we'd get the actual data back. Indeed, if I look at git blame, this check was introduced by yourself in 9134c79f4c3237297af30fe87b8e2e6c9f08e90d a few months ago, which lines up with the initial approximation of when I started seeing this bug:

https://github.com/photoprism/photoprism/blob/382c6f274886079fb1afa4351e23856b44968f01/internal/photoprism/mediafile.go#L74-L79

**Comment by lastzero** [comment]
If separate mounts cause issues, you can use a single mount for the network share. The exact location for originals and import can then be specified via config options: https://docs.photoprism.app/getting-started/config-options/#storage

**Comment by aaliddell** [comment]
Interesting, putting the import and library in a single docker mount fixes it.

Digging more into the cifs mount options, there's the [actimeo](https://manpages.debian.org/testing/cifs-utils/mount.cifs.8.en.html#actimeo=arg) option:

> The time (in seconds) that the CIFS client caches attributes of a file or directory before it requests attribute information from a server. During this period the changes that occur on the server remain undetected until the client checks the server again.

> By default, the attribute cache timeout is set to 1 second. 

In theory it's the local client that's making the file change, not the server, so I skipped this option before as it looked like it should have no effect in this situation. However, I guess multiple docker mounts make it look like there are multiple clients, so setting `actimeo=0` also fixes the bug (after restarting container to propagate mount changes).

In summary, if there are two mounts looking at the same network drive and the image file copy completes in under 1 second (default `actimeo`), the cache still says the file is zero length when it is `stat`ed and the import fails. When the manual index looks again later, the cache is expired and sees the correct value. Your single mount method is easier though and doesn't require the performance impact of disabling caching.

Thanks for the pointers :smile:. Hopefully this page helps anyone else that hits this, although it's somewhat obscure.

**Comment by lastzero** [comment]
Glad it's working now! Feel free to send a PR for our docs (FAQ or Advanced Topics section).

**Comment by aaliddell** [comment]
> Glad it's working now! Feel free to send a PR for our docs (FAQ or Advanced Topics section).

Added a pointer to this issue in https://github.com/photoprism/photoprism-docs/pull/127

**Comment by sagar5534** [comment]
I believe this issue is still apparent when using the WebDav Upload which defaults to the upload folder located in storage/users/XXX/uploads and is not a simple fix using the methods shown above.

Use Case - The user is using docker to run photoprism + connecting to an NFS/CIFS share. 
Storage Directory is on the local system or a docker volume
Originals Directory is on the network share

The solution above is the bring 2 mounts into 1 common mount within docker. The only method to achieve this is to bring the Storage directory on the NFS share as well. I can assume this is not preferable for performance issues. 

Until there is a solution to this. Uploads are broken for most users using NFS + Docker. 


PS. First time commenting here. Please let me know if more information is needed, Thanks
